host/xtest/regression_4000.c

```diff
@@ -6192,3 +6192,482 @@ static void xtest_tee_test_4016_ed25519(ADBG_Case_t *c)
 }
 ADBG_CASE_DEFINE(regression, 4016, xtest_tee_test_4016_ed25519,
 		 "Test TEE Internal API ED25519 sign/verify");
+
+struct buf_state {
+	bool buffer_two_blocks;
+	bool auth_enc;
+	bool decrypt;
+	uint8_t *plain_text;
+	uint8_t *ciph_text;
+	uint8_t *tmp_text;
+	size_t text_size;
+	size_t block_size;
+	uint8_t *iv;
+	size_t iv_len;
+	uint8_t tag[96 / 8];
+	uint8_t tmp_tag[96 / 8];
+	TEE_OperationHandle oph;
+	TEEC_Session *s;
+
+	size_t in_count;
+	size_t out_count;
+	size_t buf_count;
+};
+
+static bool process_bytes_4017(ADBG_Case_t *c, struct buf_state *bs,
+			       size_t count)
+{
+	TEE_Result res = TEEC_SUCCESS;
+	size_t new_buf_count = 0;
+	void *reference = NULL;
+	void *src = NULL;
+	size_t out = 0;
+	size_t dlen = 0;
+
+	if (bs->decrypt) {
+		src = bs->ciph_text + bs->in_count;
+		reference = bs->plain_text + bs->out_count;
+	} else {
+		src = bs->plain_text + bs->in_count;
+		reference = bs->ciph_text + bs->out_count;
+	}
+
+	bs->in_count += count;
+	if (bs->buffer_two_blocks) {
+		if (bs->in_count > bs->block_size)
+			new_buf_count = ((bs->in_count - 1) % bs->block_size) +
+					bs->block_size + 1;
+		else
+			new_buf_count = bs->in_count;
+	} else {
+		new_buf_count = bs->in_count % bs->block_size;
+	}
+
+	out = bs->buf_count + count - new_buf_count;
+	bs->out_count += out;
+	bs->buf_count = new_buf_count;
+
+	if (bs->auth_enc) {
+		res = ta_crypt_cmd_ae_update(c, bs->s, bs->oph, src, count,
+					     NULL, &dlen);
+		if (!res)
+			return ADBG_EXPECT_COMPARE_UNSIGNED(c, dlen, ==, 0) &&
+			       ADBG_EXPECT_COMPARE_UNSIGNED(c, dlen, ==, out);
```

Comment:
Do you mean
```c
return ADBG_EXPECT_COMPARE_UNSIGNED(c, dlen, ==, out);
```
?

Reply:
That works, but with a NULL output buffer it should be `0` also.